File: L6_LinkedList/CPP/Find_the_Minimum_and_Maximum_Number_of_Nodes_Between_Critical_Points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
      int count=2;
      vector<int>v;
      ListNode * temp=head->next;
      ListNode * prev=head;
      while(temp->next)
      {
          if(temp->val > prev->val && temp->val > temp->next->val)
          {
              v.push_back(count);
          }
          if(temp->val < prev->val && temp->val < temp->next->val)
          {
              v.push_back(count);
          }
          count++;
          prev=temp;
          temp=temp->next;
      }
      if(v.size()<2)
      {
          return {-1,-1};
      }
    sort(v.begin(), v.end()); // Sort the vector

    int maxDifference = v[v.size() - 1] - v[0];
    int minDifference = INT_MAX;

    for (int i = 1; i < v.size(); ++i) {
        int diff = v[i] - v[i - 1];
        minDifference = min(minDifference, diff);
    }
    return {minDifference,maxDifference};
    }
};

File: LinkedList_L3/python/Find_the_Minimum_and_Maximum_Number_of_Nodes_Between_Critical_Points.py
class Solution:
    def nodesBetweenCriticalPoints(self, head: ListNode) -> List[int]:

        ct, critPts, prev = 0, [], head.val                         # keep track of node-count, critical points
                                                                    # encountered, and previous node value
        while head.next:
																	# Tests for critical points: product < 0 if and only if
            if (prev-head.val)*(head.val-head.next.val) < 0:        #          prev < head.val > head.next.val, or 
                critPts.append(ct)                                  #          prev > head.val < head.next.val
                                                      
            prev, head = head.val, head.next                        # iterates to next node and increments node-count
            ct+= 1

        n = len(critPts)                                            # fewer than 2 nodes
        if n < 2: return [-1,-1]

        mn = min((critPts[i]-critPts[i-1] for i in range(1, n)))    # list already sorted, so min is least dist between 
        mx = critPts[-1] - critPts[0]                               # consecutive elements; max is last element - 1st element

        return [mn, mx]
